tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11457,6 +11499,34 @@ class WorkGraphVisitor : public RecursiveASTVisitor<WorkGraphVisitor> {
     }
     return true;
   }
+
+  bool VisitAttr(Attr *A) {
+    switch (A->getKind()) {
+    case attr::HLSLNodeDispatchGrid:
+    case attr::HLSLNodeMaxDispatchGrid:
+      // These attributes are only valid for Broadcasting launch nodes
+      if (!nodeLaunchType.equals_lower("broadcasting")) {
+        S.Diags.Report(A->getLocation(), diag::err_hlsl_launch_type_attr)
+          << A->getSpelling() << "Broadcasting" << A->getRange();
+        // Only output the note if the source location is valid
+        if (nodeLaunchLoc.isValid())
+          S.Diags.Report(nodeLaunchLoc, diag::note_defined_here) << "Launch type";
+      }
+      break;
+    case attr::HLSLNodeMaxRecursionDepth:
+      if (cast<HLSLNodeMaxRecursionDepthAttr>(A)->getCount() > 32)
+        S.Diags.Report(A->getLocation(), diag::err_hlsl_maxrecursiondepth_exceeded)
+          << A->getRange();
+      break;
+    case attr::HLSLNumThreads:
+      auto *numThreads = cast<HLSLNumThreadsAttr>(A);
+      if (numThreads->getX() * numThreads->getY() * numThreads->getZ() > 1024)
```

Comment:
We should be able to do this verification when we create the AST node rather than in a visitor.

Reply:
Done.